commit 49f9ed1ae8b45720478d397b01d230d808077234 (refs/changes/46/4246/4)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2018-06-18T18:12:18+00:00 (1 year, 4 months ago)
    
    PUBAPI-1411 test-cloudapi failed: "attempt to delete default network - other"
    Reviewed by: Pedro P. Candel <pedro@joyent.com>
    Approved by: Pedro P. Candel <pedro@joyent.com>

diff --git a/lib/config.js b/lib/config.js
index 904d28b..152d247 100644
--- a/lib/config.js
+++ b/lib/config.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -13,6 +13,7 @@
  */
 
 var assert = require('assert-plus');
+var backoff = require('backoff');
 var fs = require('fs');
 var http = require('http');
 var https = require('https');
@@ -93,7 +94,14 @@ function configure(options) {
 }
 
 function getAccountDcConfigFromUFDS(ufdsClient, account, datacenterName,
-    options, cb) {
+    options, retryUfdsReq, cb) {
+    if (cb === undefined) {
+        cb = retryUfdsReq;
+        retryUfdsReq = undefined;
+    }
+    assert.ok(typeof (retryUfdsReq) === 'boolean' ||
+        typeof (retryUfdsReq) === 'function' ||
+        typeof (retryUfdsReq) === 'undefined');
     assert.object(ufdsClient, 'ufdsClient');
     assert.object(account, 'account');
     assert.string(datacenterName, 'datacenterName');
@@ -107,19 +115,38 @@ function getAccountDcConfigFromUFDS(ufdsClient, account, datacenterName,
     log.info({ accountUuid: accountUuid, dcName: datacenterName },
         'Getting user config');
 
-    ufdsClient.getDcLocalConfig(accountUuid, datacenterName,
-            function _afterConfGet(err, conf) {
-        if (err) {
-            if (err.name !== 'ResourceNotFoundError') {
-                cb(err);
-                return;
-            } else {
-                conf = {}; // treat an empty object as default
+    /*
+     * We want to change the order of the arguments, because backoff expects
+     * them in a specific order (i.e. last arg should be CB, not bool).
+     */
+    function getDcLocalConfigWrapper(accountid, dcName, bool, callback) {
+        ufdsClient.getDcLocalConfig(accountid, dcName, callback, bool);
+    }
+    var call = backoff.call(getDcLocalConfigWrapper, accountUuid,
+        datacenterName, true,
+        function _afterConfGet(err, conf) {
+            if (err) {
+                if (err.name !== 'ResourceNotFoundError') {
+                    cb(err);
+                    return;
+                } else {
+                    // treat an empty object as default
+                    conf = {};
+                }
             }
+            cb(null, conf);
+        });
+    call.retryIf(function (err) {
+        var doRetry = retryUfdsReq && err.name !== 'ResourceNotFoundError';
+        if (doRetry) {
+            log.debug({ accountUuid: accountUuid, dcName: datacenterName },
+                'Retrying getDcLocalConfig');
         }
-
-        cb(null, conf);
-    }, true); // disable caching
+        return doRetry;
+    });
+    call.setStrategy(new backoff.ExponentialStrategy());
+    call.failAfter(10);
+    call.start();
     /*
      * Fetching the config is an uncommon operation, and caching the value
      * causes inconsistencies since cloudapi usually operates with multiple
@@ -134,4 +161,4 @@ module.exports = {
     configure: configure,
     getAccountDcConfigFromUFDS: getAccountDcConfigFromUFDS,
     translateUfdsConf: translateUfdsConf
-};
\ No newline at end of file
+};
diff --git a/lib/endpoints/config.js b/lib/endpoints/config.js
index 6087169..9972f9f 100644
--- a/lib/endpoints/config.js
+++ b/lib/endpoints/config.js
@@ -78,7 +78,7 @@ function getConfig(req, res, next) {
     modConfig.getAccountDcConfigFromUFDS(req.sdc.ufds, req.account,
         req.config.datacenter_name, {
         log: req.log
-    }, function _afterGetFromUFDS(err, conf) {
+    }, true, function _afterGetFromUFDS(err, conf) {
         if (err) {
             return next(translateErr(err));
         }
diff --git a/package.json b/package.json
index df1829b..e78f371 100644
--- a/package.json
+++ b/package.json
@@ -17,6 +17,7 @@
         "aperture-config": "git+https://github.com/joyent/aperture-config.git#master",
         "asn1": "0.1.11",
         "assert-plus": "1.0.0",
+        "backoff": "2.5.0",
         "bunyan": "1.8.1",
         "clone": "0.1.5",
         "ctype": "0.5.2",
